### PR TITLE
Speeding up Math.Min and Math.Max for float/double

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves https://github.com/dotnet/coreclr/issues/22951 for x86/x64.

### CODE REVIEW

**Comment by tannergooding** [comment]
For the sample code listed in the linked issue, the time went from 57ms down to 36ms (which is on-par with the netcoreapp2.2 results).

This is slightly worse than the initial prototype that saw 30ms due to the need to handle the difference between -0.0 and +0.0

File: src/System.Private.CoreLib/shared/System/Math.cs [hunk 1]
```diff
@@ -536,35 +537,91 @@ public static decimal Max(decimal val1, decimal val2)
             return decimal.Max(val1, val2);
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Math.cs]: What is the size of code that this expands to? This does not look like a good AggressiveInlining candidate...

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Math.cs]: For modern processors with AVX support (essentially any CPU from 2011 onwards), this expands to (plus or minus a couple bytes depending on registers used):

Math.Min - 32 bytes:
```asm
	vminss      xmm2, xmm1, xmm0
	vcmpeqss    xmm3, xmm1, xmm0
	vorps       xmm4, xmm1, xmm0
	vandps      xmm4, xmm4, xmm3
	vorps       xmm2, xmm2, xmm4
	vcmpunordss xmm0, xmm0, xmm0
	vblendvps   xmm0, xmm2, xmm1, xmm0
```

Math.Max - 30 bytes:
```asm
	vmaxss      xmm2, xmm1, xmm0
	vcmpeqss    xmm3, xmm1, xmm0
	vandps      xmm4, xmm1, xmm0
	vblendvps   xmm2, xmm2, xmm4, xmm3
	vcmpunordss xmm0, xmm0, xmm0
	vblendvps   xmm0, xmm2, xmm1, xmm0
```

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Math.cs]: For older processors, without AVX support but with SSE4.1 support (~2007 onwards), this expands to:

Math.Min - 52 bytes:
```asm
	movaps      xmm2, xmm1
	minss       xmm2, xmm0
	movaps      xmm3, xmm1
	cmpeqss     xmm3, xmm0
	movaps      xmm4, xmm1
	orps        xmm4, xmm0
	andps       xmm4, xmm3
	orps        xmm2, xmm4
	movaps      xmm3, xmm0
	cmpunordss  xmm3, xmm0
	movaps      xmm0, xmm3
	blendvps    xmm2, xmm1, xmm0
	movaps      xmm0, xmm2
```

Math.Max - 56 bytes:
```asm
	movaps      xmm2, xmm1
	maxss       xmm2, xmm0
	movaps      xmm3, xmm1
	cmpeqss     xmm3, xmm0
	movaps      xmmword ptr [rsp],xmm0
	movaps      xmm4, xmm1
	andps       xmm4, xmm0
	movaps      xmm0, xmm3
	blendvps    xmm2, xmm4, xmm0
	movaps      xmm0, xmmword ptr [rsp]
	movaps      xmm3, xmm0
	cmpunordss  xmm3, xmm0
	movaps      xmm0, xmm3
	blendvps    xmm2, xmm1, xmm0
	movaps      xmm0, xmm2
```

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Math.cs]: Even older x86 CPUs (anything else the runtime supports) expands to 56-bytes (`Math.Min`) and 65-bytes (`Math.Max`).

**Comment by mikedn** [comment]
At least on my machine (Haswell i5-4440) this version doesn't appear to be such an improvement. Basically I get:

Version | Time
--|--
Min20 | 38s
Min30 | 54s
Min30 + AggressiveInlinig | 45s
MinPR | 48s
MinPR - AggressiveInlining | 51s

All number obtained by actually compiling the relevant `Min` code as part of the benchmark, without using the actual corelib implementation.

File: src/System.Private.CoreLib/shared/System/Math.cs (hunk 1 above)
Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Math.cs]: The cost of not-inlining looks to be ~29 bytes of code per invocation (must save and restore the floating-point values):
```asm
movss       dword ptr [rsp+0x40], xmm0  
movss       dword ptr [rsp+0x48], xmm1  
call        <method>
vmovss  xmm0,dword ptr [rsp+0x48]
vmovss  xmm1,dword ptr [rsp+0x40]
```

Given that this is, on modern CPUs, basically the same cost as the method, and on older CPUs nearly half the cost of the actual method; it should be almost always profitable to inline (and various micro-benchmarks show the same).

**Comment by tannergooding** [comment]
> At least on my machine (Haswell i5-4440) this version doesn't appear to be such an improvement

My numbers are from my Ryzen 1800X.

Looks like I see similar results to you (an improvement from current, but not parity with 2.2) on my Kaby Lake i7-7700:
* C# netcoreapp2.2: 30sec
* C# netcoreapp3.0: 42sec
* This PR: 36sec

Tiered Compilation is left on for all of this, as it only shows minimal impact with it disabled. Mine numbers are also testing the appropriate changes to S.P.Corelib.

**Comment by tannergooding** [comment]
We could likely save some additional time if we didn't worry abort ordering -0.0 against +0.0.

The IEEE spec does not dictate this ordering and leaves it up to the implementer to decide if they should be equal or sorted:
> minNum(x, y) is the canonicalized number x if x < y, y if y < x, the canonicalized number if one
operand is a number and the other a quiet NaN. Otherwise it is either x or y, canonicalized (this
means results might differ among implementations). When either x or y is a signalingNaN, then the
result is according to 6.2.

It also has a further note under the "reproducible floating-point results" section that you shouldn't rely on which zero is returned:
> Do not depend on the sign of a zero result or the quantum of a decimal result for minNum(x, y),
maxNum(x, y), minNumMag(x, y), or maxNumMag(x, y) when x and y are equal.

That being said, the current behavior was selected because it is the behavior followed by many other implementations and is the recommended behavior by the C/C++ language standards (not that it necessarily means much with regards to C#; other than making code porting more predictable).

Realistically though, I think this is a case of do you want "compliant" code or do you want "fast" code; if we had something similar to an `/fp:fast` switch, we could decide to just optimize this down to `minss` and nothing else. This has also come up in the past with regards to automatically folding `x * y + z` into `fma` instructions and optimizing other floating-point operations.

**Comment by mikedn** [comment]
Yeah, the negative zero check is a pain in the back. I'm not quite sure why it is recommended, considering that `-0.0 < 0.0` is actually false.

Clang doesn't appear to respect this: https://gcc.godbolt.org/z/mcR-oP. As for MSVC, well, its implementation is so slow that the raytracer takes 72s instead of 32s that it takes with `l < r ? l : r`.

But perhaps it makes sense to keep the negative zero check in. People can always write their own fast version of min/max, typically just `x < y ? x : y`. And there are also min/max SSE intrinsics that they can use if they fit the bill. But it's less likely that they'll bother writing the "recommended" version so it's probably good for the framework to offer that.

**Comment by tannergooding** [comment]
Closing this in favor of https://github.com/dotnet/coreclr/pull/24039, which keeps `+0` as greater than `-0` (as will be required in IEEE 754:2019) but continues propagating NaN inputs, so we don't break WPF (this matches the new `minimum`/`maximum` functions defined in 754:2019).

**Pull request closed without merging**